cylc/flow/command_validation.py

```diff
@@ -0,0 +1,227 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Cylc command argument validation logic."""
+
+
+from typing import (
+    List,
+    Optional,
+)
+
+from cylc.flow.exceptions import InputError
+from cylc.flow.id import Tokens
+from cylc.flow.task_outputs import TASK_OUTPUT_SUCCEEDED
+from cylc.flow.flow_mgr import FLOW_ALL, FLOW_NEW, FLOW_NONE
+
+
+ERR_OPT_FLOW_VAL = "Flow values must be an integer, or 'all', 'new', or 'none'"
+ERR_OPT_FLOW_INT = "Multiple flow options must all be integer valued"
+ERR_OPT_FLOW_WAIT = (
+    f"--wait is not compatible with --flow={FLOW_NEW} or --flow={FLOW_NONE}"
+)
+
+
+def flow_opts(flows: List[str], flow_wait: bool) -> None:
+    """Check validity of flow-related CLI options.
+
+    Note the schema defaults flows to ["all"].
+
+    Examples:
+        Good:
+        >>> flow_opts(["new"], False)
+        >>> flow_opts(["1", "2"], False)
+        >>> flow_opts(["1", "2"], True)
+
+        Bad:
+        >>> flow_opts(["none", "1"], False)
+        Traceback (most recent call last):
+        cylc.flow.exceptions.InputError: ... must all be integer valued
+
+        >>> flow_opts(["cheese", "2"], True)
+        Traceback (most recent call last):
+        cylc.flow.exceptions.InputError: ... or 'all', 'new', or 'none'
+
+        >>> flow_opts(["new"], True)
+        Traceback (most recent call last):
+        cylc.flow.exceptions.InputError: ...
+
+    """
+    for val in flows:
+        val = val.strip()
+        if val in [FLOW_NONE, FLOW_NEW, FLOW_ALL]:
+            if len(flows) != 1:
+                raise InputError(ERR_OPT_FLOW_INT)
+        else:
+            try:
+                int(val)
+            except ValueError:
+                raise InputError(ERR_OPT_FLOW_VAL.format(val))
+
+    if flow_wait and flows[0] in [FLOW_NEW, FLOW_NONE]:
+        raise InputError(ERR_OPT_FLOW_WAIT)
+
+
+def prereqs(prereqs: Optional[List[str]]):
```

Comment:
Does this type hint make sense?

```suggestion
def prereqs(prereqs: Optional[List[str]] = None):
```

Reply:
Yes, Optional[str] means Union[str, None]. This argument must be provided, it may be a string or it may be None.

Optional types are often used in combination with kwargs that have None defaults, but do not have to be.

Reply:
OK - I think you're right about it being a misnomer: With this signature I'd expect

```
prereqs()
```
to work...